src/Network/Wai/SAML2/Request.hs

```diff
@@ -0,0 +1,53 @@
+module Network.Wai.SAML2.Request
+    ( Request(..)
+    , issueRequest
+    ) where
+
+import Crypto.Random
+import Data.Time.Clock
+import Data.Time.Format
+import Network.Wai.SAML2.XML
+import Text.XML
+import qualified Data.ByteString as B
+import qualified Data.ByteString.Base64 as Base64
+import qualified Data.ByteString.Lazy as BL
+import qualified Data.Map.Strict as Map
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+
+-- | Parameters for SP-initiated SSO
+data Request = Request
+  { requestIssuer :: T.Text
+  , requestAllowCreate :: Bool
+  , requestNameIDFormat :: T.Text
+  }
+
+-- | Generated a base64-encoded AuthnRequest for SP initiated SSO.
+-- See also: http://docs.oasis-open.org/security/saml/Post2.0/sstc-saml-tech-overview-2.0-cd-02.html#5.1.2.SP-Initiated%20SSO:%20%20Redirect/POST%20Bindings|outline
+issueRequest :: Request -> IO B.ByteString
+issueRequest Request{..} = do
+    now <- getCurrentTime
+    let timestamp = T.pack $ formatTime defaultTimeLocale "%Y-%m-%dT%H:%M:%SZ" now
```

Comment:
Could this be using the format from https://github.com/mbg/wai-saml2/blob/d4191b5d2d22e36746db67517b60cccf6d40f7a0/src/Network/Wai/SAML2/XML.hs#L66 to avoid the string constant here?

Reply:
fixed

Reply:
It seems that you still re-define the format in `Network.Wai.SAML2.Request`. What I meant was that you should use `timeFormat` from `Network.Wai.SAML2.XML` to avoid duplication. Sorry if that wasn't clear. I appreciate that it's not currently exported from `Network.Wai.SAML2.XML`, so the options are:

- Adding it to `Network.Wai.SAML2.XML`'s export list.
- Defining a `formatUTCTime` function in `Network.Wai.SAML2.XML` and exporting that.

When I was looking over this, I was also wondering if we could just use [the ISO8601 functions from the `time` package](https://hackage.haskell.org/package/time-1.12.2/docs/Data-Time-Format-ISO8601.html) to begin with, but I didn't find the SAML2 specification to be super clear whether it fully conforms to ISO8601.

Reply:
Oh sorry, I wasn't reading the comment carefully... I've just updated the code so that it uses `timeFormat` exported from `Network.Wai.SAML2.XML`